File: apps/ios/Lycoris/Features/Contributions/ContributionDraft.swift
import CryptoKit
import Foundation

enum ContributionFailure: Error, Equatable {
  case accountBusy, invalidReceipt, missingPhoto, storage, invalidFields
}

struct ContributionFields: Codable, Equatable {
  var title = ""
  var category = PlaceCategory.toilet
  var description = ""
  var openTimeStart = ""
  var openTimeEnd = ""
  var language = "en"
  /// The selected venue for an accessible toilet. `nil` means "not specified".
  /// Optional so pre-upgrade drafts without the key still decode.
  var venueType: PlaceVenue? = nil
  /// A raw server venue value this app does not recognize. Kept so editing other
  /// fields never silently rewrites an unknown tag to `other`.
  var unknownVenueType: String? = nil

  var valid: Bool {
    !title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
      && title.unicodeScalars.count <= 120
      && ((openTimeStart.isEmpty && openTimeEnd.isEmpty)
        || (Self.validTime(openTimeStart) && Self.validTime(openTimeEnd)))
  }

  static func validTime(_ value: String) -> Bool {
    OpeningStatusEngine.isValidTime(value)
  }

  /// A brand-new accessible toilet starts at the `other` default. Old drafts
  /// that decode without the key stay `nil` until the user chooses.
  init(language: String) {
    self.language = language
    self.venueType = .other
  }
  init(marker: Marker) {
    title = marker.title
    category = marker.category
    description = marker.description ?? ""
    openTimeStart = marker.openTimeStart ?? ""
    openTimeEnd = marker.openTimeEnd ?? ""
    language = marker.contentLanguage
    venueType = marker.venue
    // Only an accessible toilet may carry a venue; ignore a stale server value
    // on any other category so it can never be resent.
    unknownVenueType =
      marker.category == .toilet && marker.venue == nil ? marker.venueType : nil
  }

  // Explicit Codable so a pre-upgrade draft JSON (without the new keys) decodes
  // with defaults instead of failing the whole journal.
  private enum CodingKeys: String, CodingKey {
    case title, category, description, openTimeStart, openTimeEnd, language
    case venueType, unknownVenueType
  }

  init(from decoder: any Decoder) throws {
    let container = try decoder.container(keyedBy: CodingKeys.self)
    title = try container.decodeIfPresent(String.self, forKey: .title) ?? ""
    category =
      try container.decodeIfPresent(PlaceCategory.self, forKey: .category) ?? .toilet
    description = try container.decodeIfPresent(String.self, forKey: .description) ?? ""
    openTimeStart = try container.decodeIfPresent(String.self, forKey: .openTimeStart) ?? ""
    openTimeEnd = try container.decodeIfPresent(String.self, forKey: .openTimeEnd) ?? ""
    language = try container.decodeIfPresent(String.self, forKey: .language) ?? "en"
    venueType = try? container.decodeIfPresent(PlaceVenue.self, forKey: .venueType)
    unknownVenueType = try container.decodeIfPresent(String.self, forKey: .unknownVenueType)
  }
}

struct ContributionDraft: Codable, Equatable, Identifiable {
  enum Phase: String, Codable { case draft, creating, editing, uncertainEdit, uploading, complete }
  let id: UUID
  let owner: String
  let origin: String
  var point: GeoPoint
  let original: Marker?
  var fields: ContributionFields
  var phase = Phase.draft
  var requestBody: Data?
  var markerID: Int64?
  var photoID: UUID?
  var photoHash: String?
  var photoSize: Int?
  var upload: UploadReceipt?
  var photoRejected = false

  init(owner: String, origin: String, point: GeoPoint, language: String, marker: Marker? = nil) {
    id = UUID()
    self.owner = owner
    self.origin = origin
    self.point = point
    original = marker
    fields = marker.map(ContributionFields.init) ?? ContributionFields(language: language)
    markerID = marker?.id
  }

  var editable: Bool { phase == .draft }
  var hasChanges: Bool { original.map { fields != ContributionFields(marker: $0) } ?? true }
  var canSubmit: Bool { fields.valid && (hasChanges || photoID != nil) }

  var validCheckpoint: Bool {
    guard !owner.isEmpty, !origin.isEmpty,
      GeoPoint(latitude: point.latitude, longitude: point.longitude) != nil,
      markerID == nil || markerID! > 0
    else { return false }
    if [.uploading, .complete].contains(phase), markerID == nil { return false }
    if [.creating, .editing, .uncertainEdit].contains(phase), requestBody == nil { return false }
    if [.editing, .uncertainEdit].contains(phase), original == nil { return false }
    if phase == .creating && original != nil { return false }
    if let original, original.point != point || original.id != markerID { return false }
    if let photoSize, let photoHash, photoID != nil {
      guard (1...(5 * 1024 * 1024)).contains(photoSize),
        photoHash.range(of: #"^[a-f0-9]{64}$"#, options: .regularExpression) != nil
      else { return false }
    } else if photoID != nil || photoSize != nil || photoHash != nil || phase == .uploading {
      return false
    }
    if let upload { return (try? upload.validate(for: self)) != nil }
    return true
  }

  func encodedRequest() throws -> Data {
    var json: [String: Any] = [
      "title": fields.title, "description": fields.description,
      "category": fields.category.rawValue, "language": fields.language,
      "openTimeStart": fields.openTimeStart, "openTimeEnd": fields.openTimeEnd,
    ]
    // Only accessible toilets may carry a venue, and the tag must never be sent
    // for another category (the server clears it then). A new toilet that keeps
    // the default `other` omits the field, which the server defaults to `other`.
    // Only accessible toilets may carry a venue; never send a stale toilet tag
    // for another category (the server clears it then).
    if fields.category == .toilet, let venue = fields.venueType {
      // A new toilet keeping the default `other` omits the field, which the
      // server defaults to `other`; an edit always sends the current value so a
      // deliberate change (including back to `other`) is not lost.
      // An unrecognized server value (venueType == nil) is omitted so PATCH
      // preserves the server's original instead of resending an unknown raw to
      // the six-value validator.
      if original != nil || venue != .other {
        json["venueType"] = venue.rawValue
      }
    }
    if original == nil {
      json["lat"] = point.latitude
      json["lng"] = point.longitude
      json["clientRequestId"] = id.uuidString
    }
    return try JSONSerialization.data(withJSONObject: json, options: .sortedKeys)
  }
}

struct UploadReceipt: Codable, Equatable {
  let uploadId: String
  let markerId: Int64
  let totalBytes: Int
  let receivedBytes: Int
  let chunkSize: Int
  let status: String
  var complete: Bool { status == "COMPLETED" }

  func validate(for draft: ContributionDraft) throws {
    guard UUID(uuidString: uploadId) != nil, markerId == draft.markerID,
      totalBytes == draft.photoSize, chunkSize == 262_144,
      receivedBytes >= 0, receivedBytes <= totalBytes,
      receivedBytes == totalBytes || receivedBytes % chunkSize == 0,
      ["UPLOADING", "COMPLETED"].contains(status),
      !complete || receivedBytes == totalBytes,
      draft.upload == nil || draft.upload?.uploadId == uploadId,
      receivedBytes >= (draft.upload?.receivedBytes ?? 0)
    else { throw ContributionFailure.invalidReceipt }
  }
}

/// One private, atomic journal. The final encoded image is written first, never
/// regenerated on retry. Neither file participates in device/iCloud backups.
struct ContributionJournal {
  let directory: URL
  init(directory: URL? = nil) {
    self.directory =
      directory
      ?? URL.applicationSupportDirectory.appendingPathComponent(
        "Contribution", isDirectory: true)
  }
  private var record: URL { directory.appendingPathComponent("draft.json") }
  func photoURL(_ id: UUID) -> URL { directory.appendingPathComponent("\(id.uuidString).jpg") }

  func load() throws -> ContributionDraft? {
    guard FileManager.default.fileExists(atPath: record.path) else { return nil }
    let value = try JSONDecoder().decode(ContributionDraft.self, from: Data(contentsOf: record))
    guard value.validCheckpoint else { throw ContributionFailure.storage }
    return value
  }
  func save(_ draft: ContributionDraft) throws {
    try prepare()
    try JSONEncoder().encode(draft).write(
      to: record, options: [.atomic, .completeFileProtectionUntilFirstUserAuthentication])
  }
  func savePhoto(_ data: Data, id: UUID) throws -> (hash: String, size: Int) {
    guard !data.isEmpty, data.count <= 5 * 1024 * 1024 else { throw AccountFailure(status: 413) }
    try prepare()
    try data.write(
      to: photoURL(id), options: [.atomic, .completeFileProtectionUntilFirstUserAuthentication])
    // Hash the durable bytes, exactly as subsequent requests will read them.
    let saved = try Data(contentsOf: photoURL(id))
    return (Self.hash(saved), saved.count)
  }
  func photo(_ draft: ContributionDraft) throws -> Data {
    guard let id = draft.photoID, let data = try? Data(contentsOf: photoURL(id)),
      data.count == draft.photoSize, Self.hash(data) == draft.photoHash
    else { throw ContributionFailure.missingPhoto }
    return data
  }
  func removePhoto(_ id: UUID?) {
    if let id { try? FileManager.default.removeItem(at: photoURL(id)) }
  }
  func clear() throws {
    if FileManager.default.fileExists(atPath: directory.path) {
      try FileManager.default.removeItem(at: directory)
    }
  }
  private func prepare() throws {
    try FileManager.default.createDirectory(
      at: directory, withIntermediateDirectories: true,
      attributes: [
        .posixPermissions: 0o700,
        .protectionKey: FileProtectionType.completeUntilFirstUserAuthentication,
      ])
    var location = directory
    var values = URLResourceValues()
    values.isExcludedFromBackup = true
    try location.setResourceValues(values)
  }
  static func hash(_ data: Data) -> String {
    SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
  }
}

File: apps/android/app/src/main/java/com/lycoris/maps/feature/contributions/ContributionDraft.kt
package com.lycoris.maps.feature.contributions

import com.lycoris.maps.core.media.EncodedPhoto
import com.lycoris.maps.core.media.PhotoPolicy
import com.lycoris.maps.core.media.canonicalUuid
import com.lycoris.maps.core.model.Marker
import com.lycoris.maps.core.model.PlaceCategory
import com.lycoris.maps.core.model.validCoordinate
import com.lycoris.maps.core.network.CreateMarkerRequest
import com.lycoris.maps.core.network.EditMarkerRequest
import com.lycoris.maps.core.network.LycorisJson
import com.lycoris.maps.core.network.UploadReceipt
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString

@Serializable
data class ContributionFields(
    val title: String = "",
    val category: String = PlaceCategory.ACCESSIBLE_TOILET.wireValue,
    val description: String = "",
    val openTimeStart: String = "",
    val openTimeEnd: String = "",
    val language: String = "en",
) {
    fun isValid(): Boolean = title.trim().isNotEmpty() && title.trim().let { it.codePointCount(0, it.length) <= 120 } &&
        category in PlaceCategory.entries.map { it.wireValue } && language in setOf("en", "zh") &&
        ((openTimeStart.isEmpty() && openTimeEnd.isEmpty()) || (validTime(openTimeStart) && validTime(openTimeEnd)))

    companion object {
        fun fromMarker(marker: Marker): ContributionFields = ContributionFields(
            marker.title, marker.category, marker.description.orEmpty(), marker.openTimeStart.orEmpty(),
            marker.openTimeEnd.orEmpty(), marker.contentLanguage,
        )
        private fun validTime(value: String) = value.matches(Regex("(?:[01][0-9]|2[0-3]):[0-5][0-9]"))
    }
}

@Serializable
enum class DraftPhase { DRAFT, CREATING, EDITING, UNCERTAIN_EDIT, UPLOADING, COMPLETE }
@Serializable
enum class DraftProblem { SESSION_REQUIRED, NETWORK, RETRY_LIMIT, INVALID_RECEIPT, INVALID_FIELDS, MISSING_PHOTO, PHOTO_EXPIRED, PHOTO_REJECTED, INACCESSIBLE, CONFLICT, STORAGE, UNCERTAIN_EDIT }

@Serializable
data class ContributionDraft(
    val id: String,
    val owner: String,
    val origin: String,
    val latitude: Double,
    val longitude: Double,
    val fields: ContributionFields,
    val original: Marker? = null,
    val creationRequestId: String = id,
    val phase: DraftPhase = DraftPhase.DRAFT,
    val frozenRequest: String? = null,
    val markerId: Long? = original?.id,
    val photo: EncodedPhoto? = null,
    val upload: UploadReceipt? = null,
    val problem: DraftProblem? = null,
    val paused: Boolean = false,
    val attempts: Int = 0,
    val revision: Long = 0,
    val updatedAt: Long = 0,
) {
    val editable: Boolean get() = phase == DraftPhase.DRAFT
    val hasTextChanges: Boolean get() = original == null || fields != ContributionFields.fromMarker(original)
    val canSubmit: Boolean get() = editable && fields.isValid() && (hasTextChanges || photo != null)
    val safelyResumable: Boolean get() = phase == DraftPhase.CREATING || phase == DraftPhase.UPLOADING
    val canReplacePhoto: Boolean get() = editable || (phase == DraftPhase.UPLOADING &&
        problem in setOf(DraftProblem.PHOTO_EXPIRED, DraftProblem.PHOTO_REJECTED, DraftProblem.MISSING_PHOTO))

    fun frozenBody(): String {
        require(canSubmit)
        return if (original == null) LycorisJson.encodeToString(CreateMarkerRequest(
            latitude, longitude, fields.category, fields.title.trim(), fields.description, fields.language,
            fields.openTimeStart, fields.openTimeEnd, creationRequestId,
        )) else LycorisJson.encodeToString(EditMarkerRequest(
            fields.category, fields.title.trim(), fields.description, fields.language, fields.openTimeStart, fields.openTimeEnd,
        ))
    }

    fun isValidCheckpoint(): Boolean {
        if (canonicalUuid(id) != id || canonicalUuid(creationRequestId) != creationRequestId || owner.isBlank() || origin.isBlank() || !validCoordinate(latitude, longitude) ||
            markerId?.let { it <= 0L } == true || revision < 0 || attempts < 0) return false
        if (original != null && (markerId != original.id || latitude != original.lat || longitude != original.lng)) return false
        if (phase in setOf(DraftPhase.CREATING, DraftPhase.EDITING, DraftPhase.UNCERTAIN_EDIT) && frozenRequest.isNullOrBlank()) return false
        if (phase == DraftPhase.CREATING && original != null) return false
        if (phase in setOf(DraftPhase.EDITING, DraftPhase.UNCERTAIN_EDIT) && original == null) return false
        if (phase in setOf(DraftPhase.UPLOADING, DraftPhase.COMPLETE) && markerId == null) return false
        if (phase == DraftPhase.UPLOADING && photo == null) return false
        if (photo != null && !photo.isValid()) return false
        if (upload != null && !UploadReceiptPolicy.valid(upload, this, previous = null)) return false
        return true
    }
}

object UploadReceiptPolicy {
    fun valid(receipt: UploadReceipt, draft: ContributionDraft, previous: UploadReceipt? = draft.upload): Boolean {
        val photo = draft.photo ?: return false
        val normalizedId = canonicalUuid(receipt.uploadId) ?: return false
        return receipt.markerId == draft.markerId && receipt.totalBytes == photo.byteCount &&
            receipt.chunkSize == PhotoPolicy.UPLOAD_CHUNK_BYTES && receipt.receivedBytes in 0..receipt.totalBytes &&
            (receipt.receivedBytes == receipt.totalBytes || receipt.receivedBytes % receipt.chunkSize == 0) &&
            receipt.status in setOf("UPLOADING", "COMPLETED") &&
            (receipt.status != "COMPLETED" || receipt.receivedBytes == receipt.totalBytes) &&
            (previous == null || (normalizedId == canonicalUuid(previous.uploadId) &&
                receipt.receivedBytes >= previous.receivedBytes &&
                (previous.status != "COMPLETED" || receipt.status == "COMPLETED")))
    }
}

class DraftStorageFailure : Exception("Could not save contribution draft")
class InvalidUploadReceipt : Exception("Upload receipt does not match this draft")
